utils/rate_limiter.go

```diff
@@ -47,18 +49,42 @@ type rateLimiter struct {
 //
 // It can also be used to limit the rate of traffic in bytes, by setting creditsPerSecond to desired throughput
 // as bytes/second, and calling CheckCredit() with the actual message size.
-func NewRateLimiter(creditsPerSecond, maxBalance float64) RateLimiter {
-	return &rateLimiter{
+func NewRateLimiter(creditsPerSecond, maxBalance float64) *RateLimiterImpl {
+	return &RateLimiterImpl{
 		creditsPerSecond: creditsPerSecond,
-		balance:          maxBalance,
+		balance:          maxBalance * rand.Float64(),
 		maxBalance:       maxBalance,
 		lastTick:         time.Now(),
-		timeNow:          time.Now}
+		timeNow:          time.Now,
+	}
+}
+
+// Update updates the rate limiter with new configurations. This allows the balance to carry over as the rate changes.
+func (b *RateLimiterImpl) Update(creditsPerSecond, maxBalance float64) {
+	b.lock.Lock()
+	defer b.lock.Unlock()
+	b.updateBalance()
+	b.creditsPerSecond = creditsPerSecond
+	// The new balance should be proportional to the old balance.
+	b.balance = maxBalance * b.balance / b.maxBalance
+	b.maxBalance = maxBalance
 }
 
-func (b *rateLimiter) CheckCredit(itemCost float64) bool {
-	b.Lock()
-	defer b.Unlock()
+// CheckCredit returns true if the RateLimiter has enough credit balance for the itemCost.
+func (b *RateLimiterImpl) CheckCredit(itemCost float64) bool {
```

Comment:
I think we discussed elsewhere that we should really accept the timestamp here, since it's always available when starting the span. A minor perf optimization for the future. But can wait if we move this into `internal`